Design.MD

```diff
@@ -0,0 +1,85 @@
+## Use cases
+### Async programming
+### Generators
+## Implementation
+### Api
+#### Public
+
+```scala
+trait Continuation[A]
+  val context: ExecutionContext
```

Comment:
A `Continuation` runs on a single "`Thread`" (`ExecutorService`/Context/pool), but it can interleave many other `Continuation`s running on other threads/pools.

So this `val context` represents the context this `Continuation` is running on,
if you run a `suspend () -> A` on a different context this requires you to instantiate a new `Continuation` with a new `val context`. This is useful so you can rely on `Continuation` to _preserve the context_.

Context preservation is something that we're discussing in our other comments, https://github.com/47deg/TBD/pull/36#discussion_r941384915. It's basically to prevent following scenarios.

```kotlin
val x: Suspend ?=> Int = suspendCoroutine { cont ->
  Thread({
    cont.resume(1) // resumes callback from `thread-1`
  }, "thread-1").start()
}

def main: Unit =
  println(Thread.currentThread().name) // global-context
  x()
  println(Thread.currentThread().name) // thread-1
```

Due to not preserving the context correctly within the `suspendCoroutine`, we are returned not on the caller thread but on `thread-1`.

This is unrelated to offering a optimised set of dispatchers like you say, which would be a great thing to have in the library. Wondering what your thoughts are after clarifying the reason.

Additional questions from my side:

 - Can we use `ExecutorService` in Scala JS?
 - Why does Scala have `ExecutionContext` ? (Especially if previous question was answered "yes").
 - Are there effect systems using `ExecutorService` rather than `ExecutionContext`?
 
 
IIRC CE2 relied on `ExecutionContext`, even `global` but also implemented a couple of their own. I.e. for trampolining / stack-safety.